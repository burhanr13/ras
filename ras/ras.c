#include "ras.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/mman.h>

#define MASK(b) ((1 << (b)) - 1)
#define ISNBITSU(n, b) ((u32) (n) >> (b) == 0)
#define ISNBITSS(n, b) ((s32) (n) >> (b) == 0 || (s32) (n) >> (b) == -1)
#define ISLOWBITS0(n, b) (((n) & MASK(b)) == 0)

typedef enum {
    SYM_UNDEFINED,
    SYM_INTERNAL,
    SYM_EXTERNAL,
} rasSymbolType;

typedef struct _rasSymbol {
    rasSymbolType type;
    union {
        size_t intOffset; // in words
        void* extAddr;
    };
} rasSymbol;

typedef struct _rasPatch {
    rasPatchType type;
    size_t offset; // in words
    rasLabel sym;
} rasPatch;

#define LISTNODELEN 64

#define LISTNODE(T)                                                            \
    struct ListNode_##T {                                                      \
        T d[LISTNODELEN];                                                      \
        size_t count;                                                          \
        struct ListNode_##T* next;                                             \
    }*

#define LISTNEXT(l)                                                            \
    ({                                                                         \
        if (!(l) || (l)->count == LISTNODELEN) {                               \
            typeof(l) n = malloc(sizeof *n);                                   \
            n->count = 0;                                                      \
            n->next = (l);                                                     \
            (l) = n;                                                           \
        }                                                                      \
        &(l)->d[(l)->count++];                                                 \
    })

#define LISTPOP(l)                                                             \
    ({                                                                         \
        typeof(l) tmp = (l)->next;                                             \
        free(l);                                                               \
        (l) = tmp;                                                             \
    })

typedef struct _rasBlock {

    u32* code;
    u32* curr;
    size_t size; // in words

    size_t initialSize;

    LISTNODE(rasSymbol) symbols;
    LISTNODE(rasPatch) patches;

} rasBlock;

char* rasErrorStrings[] = {
    "no error",
    "register size mismatch",
    "immediate out of range",
    "invalid constant operand (shift, extend, etc)",
    "undefined label",
    "label out of range or misaligned",
};

rasErrorCallback errorCallback = NULL;

static void* jit_alloc(size_t size) {
    // try to map near the static code
    void* ptr = mmap(rasErrorStrings, size, PROT_READ | PROT_WRITE,
                     MAP_PRIVATE | MAP_ANON, -1, 0);
    if (ptr == MAP_FAILED) {
        perror("mmap");
        abort();
    }
    return ptr;
}

enum Perm {
    RW,
    RX,
    RWX,
};

static void jit_protect(void* code, size_t size, enum Perm perm) {
    switch (perm) {
        case RW:
            mprotect(code, size, PROT_READ | PROT_WRITE);
            break;
        case RX:
            mprotect(code, size, PROT_READ | PROT_EXEC);
            break;
        case RWX:
            mprotect(code, size, PROT_READ | PROT_WRITE | PROT_EXEC);
            break;
    }
}

static void jit_free(void* code, size_t size) {
    munmap(code, size);
}

static void jit_clearcache(void* code, size_t size) {
    __builtin___clear_cache(code, code + size);
}

void rasSetErrorCallback(rasErrorCallback cb) {
    errorCallback = cb;
}

rasBlock* rasCreate(size_t initialSize) {
    rasBlock* ctx = calloc(1, sizeof *ctx);

    ctx->code = jit_alloc(initialSize);
    ctx->curr = ctx->code;
    ctx->size = initialSize / 4;

    ctx->initialSize = initialSize;

    ctx->symbols = NULL;
    ctx->patches = NULL;

    return ctx;
}

void rasDestroy(rasBlock* ctx) {
    jit_free(ctx->code, 4 * ctx->size);

    free(ctx);
}

rasLabel rasDeclareLabel(rasBlock* ctx) {
    rasSymbol* l = LISTNEXT(ctx->symbols);
    l->type = SYM_UNDEFINED;
    return l;
}

rasLabel rasDefineLabel(rasBlock* ctx, rasLabel l) {
    l->type = SYM_INTERNAL;
    l->intOffset = ctx->curr - ctx->code;
    return l;
}

rasLabel rasDefineLabelExternal(rasLabel l, void* addr) {
    l->type = SYM_EXTERNAL;
    l->extAddr = addr;
    return l;
}

void* rasGetLabelAddr(rasBlock* ctx, rasLabel l) {
    switch (l->type) {
        case SYM_INTERNAL:
            return ctx->code + l->intOffset;
        case SYM_EXTERNAL:
            return l->extAddr;
        case SYM_UNDEFINED:
        default:
            return NULL;
    }
}

void rasAddPatch(rasBlock* ctx, rasPatchType type, rasLabel l) {
    rasPatch* p = LISTNEXT(ctx->patches);
    p->type = type;
    p->offset = ctx->curr - ctx->code;
    p->sym = l;
}

void rasApplyPatch(rasBlock* ctx, rasPatch p) {
    void* patchaddr = &ctx->code[p.offset];
    void* symaddr = rasGetLabelAddr(ctx, p.sym);
    rasAssert(symaddr != NULL, RAS_ERR_UNDEF_LABEL);

    ptrdiff_t reladdr = symaddr - patchaddr;

    u32* patchinst = patchaddr;

    switch (p.type) {
        case RAS_PATCH_ABSADDR: {
            *(void**) patchaddr = symaddr;
            break;
        }
        case RAS_PATCH_BRANCH26: {
            rasAssert(ISLOWBITS0(reladdr, 2), RAS_ERR_BAD_LABEL);
            reladdr >>= 2;
            rasAssert(ISNBITSS(reladdr, 26), RAS_ERR_BAD_LABEL);
            reladdr &= MASK(26);
            *patchinst |= reladdr;
            break;
        }
        case RAS_PATCH_BRANCH19: {
            rasAssert(ISLOWBITS0(reladdr, 2), RAS_ERR_BAD_LABEL);
            reladdr >>= 2;
            rasAssert(ISNBITSS(reladdr, 19), RAS_ERR_BAD_LABEL);
            reladdr &= MASK(19);
            *patchinst |= reladdr << 5;
            break;
        }
    }
}

void rasApplyAllPatches(rasBlock* ctx) {
    while (ctx->patches) {
        for (int i = 0; i < ctx->patches->count; i++) {
            rasApplyPatch(ctx, ctx->patches->d[i]);
        }
        LISTPOP(ctx->patches);
    }
}

void rasReady(rasBlock* ctx) {
    rasApplyAllPatches(ctx);

    jit_protect(ctx->code, 4 * ctx->size, RX);
    jit_clearcache(ctx->code, 4 * ctx->size);
}

void rasUnready(rasBlock* ctx) {
    jit_protect(ctx->code, 4 * ctx->size, RW);
}

void* rasGetCode(rasBlock* ctx) {
    return ctx->code;
}

size_t rasGetSize(rasBlock* ctx) {
    return (ctx->curr - ctx->code) * 4;
}

void rasThrow(rasError err) {
    if (errorCallback) {
        errorCallback(err);
    } else {
        fprintf(stderr, "ras error: %s\n", rasErrorStrings[err]);
        exit(1);
    }
}

void rasAssert(int condition, rasError err) {
    if (!condition) {
        rasThrow(err);
    }
}

static void ras_grow(rasBlock* ctx) {
    u32* oldCode = ctx->code;
    size_t oldSize = ctx->size;
    ctx->size *= 2;
    ctx->code = jit_alloc(4 * ctx->size);
    ctx->curr = ctx->code;
    memcpy(ctx->code, oldCode, 4 * oldSize);
    jit_free(oldCode, 4 * oldSize);
}

void rasEmitWord(rasBlock* ctx, u32 w) {
    if (ctx->curr == ctx->code + ctx->size) ras_grow(ctx);
    *ctx->curr++ = w;
}

void rasEmitDword(rasBlock* ctx, u64 d) {
    if ((ctx->curr - ctx->code) & 1) rasEmitWord(ctx, 0);
    rasEmitWord(ctx, d);
    rasEmitWord(ctx, d >> 32);
}
