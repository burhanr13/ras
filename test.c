#include <stdio.h>
#include <stdlib.h>

#include "ras/ras.h"
#include "ras/ras_macros.h"

int test() {
    return 2;
}

int main() {
    rasBlock* ctx = rasCreate(16384);

    str(lr, pre_ptr(sp, -0x10));
    bl(Lnew(test));
    ldr(lr, post_ptr(sp, 0x10));
    word(0xd65f03c0); // ret

    b(eq, Lnew());
    b(Lnew());

    rasReady(ctx);

    int (*f)() = rasGetCode(ctx);
    printf("%d\n", f());

    rasDestroy(ctx);
}