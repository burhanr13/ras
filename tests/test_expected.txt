b #8
bl #0
b.eq #0
b.ne #8
b.hs #0x28
b.lo #0x48
b.vs #0
b.vc #8
b.mi #0x28
b.pl #0x48
b.hi #0
b.ls #8
b.ge #0x28
b.lt #0x48
b.gt #0
b.le #8
b.al #0x28
b.nv #0x48
ldr w0, #0x5c
ldr x0, #0
ldrsw x0, #8
adr x0, #0x48
adr x0, #0x74
adrp x0, #0
add x0, x0, #0x74
cbz w0, #0x74
cbnz w0, #0x5c
cbz x0, #0x48
cbnz x0, #0x28
add w0, w1, #0x123
adds x2, x3, #1, lsl #12
sub w4, w5, #0xabc
subs x6, x7, #0xa, lsl #12
add w8, w9, w10
adds x11, x12, w13, uxtb
sub w14, w15, w16, uxth #3
subs w17, w18, w19, uxtw #3
add x20, x21, x22, uxtx #3
adds x23, x24, w25, sxtb #2
sub w26, w27, w28, sxth
subs x29, x30, w0, sxtw #1
add x0, x1, x2, sxtx #2
adds w0, w1, w2, lsl #3
sub x0, x1, x2, lsr #50
subs w0, w1, w2, asr #20
add sp, sp, #4
add w0, wsp, w1
add sp, sp, w0, sxtw #2
sub x0, sp, #4
cmp w0, w1
cmn x0, x1
cmp w0, w1, asr #5
cmp x0, x1, sxtx #3
add x0, sp, w1, uxtw #3
cmp sp, x1
cmp x0, #0
cmp w0, #0x64
sub w0, w1, #1
sub w0, w1, #0xccc, lsl #12
cmn w0, #0xccc, lsl #12
mov w2, #0x5678
movk w2, #0x1234, lsl #16
adds w0, w1, w2
adc w0, w1, w2
adcs w0, w1, w2
adc x0, x1, x2
adcs x0, x1, x2
sbc w0, w1, w2
sbcs w0, w1, w2
sbc x0, x1, x2
sbcs x0, x1, x2
and w0, w1, w2
bic x0, x1, x2, lsl #10
orr w0, w1, w2, lsr #10
orn w0, w1, w2, asr #10
eor x0, x1, x2
eon x0, x1, x2, asr #50
ands w0, w1, w2, lsl #20
bics x0, x1, x2, lsr #40
tst w0, w1
and w0, w1, #1
orr x0, x1, #0xffffffff
eor w0, w1, #0xcccccccc
ands w0, w1, #0xdfdfdfdf
tst x0, #-0x380038003800381
and w0, w1, #0xfffff000
orr x0, x1, #0xffff0000ffff00
eor w0, w1, #0xff00ff
ands w0, w1, #0xffff7fff
tst x0, #0xfff800
and sp, x0, #0xfffffffffffffff0
mov w2, #-0x30000000
and w0, w1, w2
mov w2, #0xd0f0
movk w2, #0xabc, lsl #16
orr w0, w1, w2
rbit w0, w1
rbit x0, x1
rev w0, w1
rev x0, x1
rev16 w0, w1
rev16 x0, x1
rev32 x0, x1
clz w0, w1
clz x0, x1
cls w0, w1
cls x0, x1
udiv w0, w1, w2
sdiv w0, w1, w2
lsl w0, w1, w2
lsr w0, w1, w2
asr w0, w1, w2
ror w0, w1, w2
udiv x0, x1, x2
sdiv x0, x1, x2
lsl x0, x1, x2
lsr x0, x1, x2
asr x0, x1, x2
ror x0, x1, x2
mul w0, w1, w2
mul x0, x1, x2
smull x0, w1, w2
umull x0, w1, w2
cset w0, eq
cset x0, eq
csel w0, w1, w0, eq
cinc x0, x1, eq
cinv w0, w1, eq
cneg x0, x1, eq
sbfiz w0, w1, #3, #0xa
sbfx w0, w1, #0xa, #0x14
bfi x0, x1, #0x28, #0xa
ubfx x0, x1, #0x28, #0xa
ubfiz w0, w1, #0xf, #5
lsl w0, w1, #0xa
lsr w0, w1, #0xa
asr w0, w1, #0xa
ror w0, w1, #0xa
lsl x0, x1, #0x28
lsr x0, x1, #0x28
asr x0, x1, #0x28
ror x0, x1, #0x28
lsl w0, w1, w2
ror x0, x1, x2
uxtb w0, w1
sxtb w0, w1
sxtb x0, w1
uxth w0, w1
sxth w0, w1
sxth x0, w1
sxtw x0, w1
extr w0, w1, w2, #0x14
extr x0, x1, x2, #0x32
movk w0, #0x1234
mov w0, #0x12340000
mov x0, #-0x1235
movk x0, #0x1234, lsl #16
mov x0, #0x123400000000
mov x0, #-0x1234000000000001
mov w0, #0
mov x0, #0
mov w0, #-1
mov x0, #-1
mov w0, #0x80
mov w0, #-0x80
mov w0, #-0x33333334
mov x0, #0x800
mov x0, #-0x800
mov w0, #0x5678
movk w0, #0x1234, lsl #16
mov w0, #0xabcd
mov w0, #-0x54330000
mov w0, #-0x5433
mov x0, #-0x64
mov x0, #0xcdef
movk x0, #0x78ab, lsl #16
movk x0, #0x3456, lsl #32
movk x0, #0x12, lsl #48
mov x0, #-0x5433
movk x0, #0xff, lsl #48
mov x0, #-0x5433
mov w0, w1
mov x0, x1
mov sp, x0
mov x0, sp
strb w0, [x1, x2]
ldrb w0, [x1, #0x10]
ldursb w0, [x1, #-0x10]
ldrsb x0, [x1, #0xff]!
strh w0, [x1, #0x1ffe]
ldrh w0, [x1, #0x20]!
ldrsh w0, [x1], #0xffffffffffffffe0
ldrsh x0, [x1], #0xffffffffffffff00
ldr w0, [x1, #0x3ffc]
ldrsw x0, [x1, x2, sxtx]
ldr x0, [x1, #0x7ff8]
str w0, [x1, w2, uxtw]
str x0, [x1, w2, sxtw #3]
ldr w0, [x1, x2, lsl #2]
ldrh w0, [x1, w2, uxtw #1]
str s0, [x1, x2, lsl #2]
ldr s0, [x1, w2, sxtw #2]
str q0, [x0, #0xfff0]
str q0, [x0, x1, lsl #4]
ldr q0, [x0, #0xfff0]
str q0, [sp, #-0x10]!
ldr q0, [sp], #0x10
ldp w0, w1, [x2]
stp w0, w1, [x2, #4]
ldpsw x0, x1, [x2, #0x80]!
ldp x0, x1, [x2], #-0x40
stp x0, x1, [x2, #-0x1f8]
stp x29, x30, [sp, #-0x10]!
ldp x29, x30, [sp], #0x10
stp q0, q1, [x0], #0x20
ldp q0, q1, [x0], #0x20
br x16
blr x16
ret 
ret x17
nop 
mrs x0, nzcv
fmov s0, #1.00000000
fmov s0, #0.25000000
fmov d0, #-0.25000000
fmov s0, #0.50000000
fmov d0, #-0.75000000
fmov s0, s1
fmov d0, d1
fmov s0, w0
fmov w0, s0
fmov d0, x0
fmov x0, d0
ucvtf s0, w0
fcvtzs x0, s0
fabs d0, d1
fneg s0, s1
fcvt s0, d1
fcvt d0, s1
fcmp s0, s1
fcmp d0, d1
fcmp s0, #0.0
fcmp d0, #0.0
fadd s0, s1, s2
fsub d0, d1, d2
fmul s0, s1, s2
fnmul d0, d1, d2
fdiv s0, s1, s2
fmax s0, s1, s2
fmin d0, d1, d2
fmadd s0, s1, s2, s3
fmsub d0, d1, d2, d3
fnmadd d0, d1, d2, d3
fnmsub s0, s1, s2, s3
shadd v0.8b, v1.8b, v2.8b
shadd v0.16b, v1.16b, v2.16b
shadd v0.4h, v1.4h, v2.4h
shadd v0.8h, v1.8h, v2.8h
shadd v0.2s, v1.2s, v2.2s
shadd v0.4s, v1.4s, v2.4s
sqadd v0.8b, v1.8b, v2.8b
sqadd v0.16b, v1.16b, v2.16b
sqadd v0.4h, v1.4h, v2.4h
sqadd v0.8h, v1.8h, v2.8h
sqadd v0.2s, v1.2s, v2.2s
sqadd v0.4s, v1.4s, v2.4s
srhadd v0.8b, v1.8b, v2.8b
srhadd v0.16b, v1.16b, v2.16b
srhadd v0.4h, v1.4h, v2.4h
srhadd v0.8h, v1.8h, v2.8h
srhadd v0.2s, v1.2s, v2.2s
srhadd v0.4s, v1.4s, v2.4s
shsub v0.8b, v1.8b, v2.8b
shsub v0.16b, v1.16b, v2.16b
shsub v0.4h, v1.4h, v2.4h
shsub v0.8h, v1.8h, v2.8h
shsub v0.2s, v1.2s, v2.2s
shsub v0.4s, v1.4s, v2.4s
sqsub v0.8b, v1.8b, v2.8b
sqsub v0.16b, v1.16b, v2.16b
sqsub v0.4h, v1.4h, v2.4h
sqsub v0.8h, v1.8h, v2.8h
sqsub v0.2s, v1.2s, v2.2s
sqsub v0.4s, v1.4s, v2.4s
cmgt v0.8b, v1.8b, v2.8b
cmgt v0.16b, v1.16b, v2.16b
cmgt v0.4h, v1.4h, v2.4h
cmgt v0.8h, v1.8h, v2.8h
cmgt v0.2s, v1.2s, v2.2s
cmgt v0.4s, v1.4s, v2.4s
cmge v0.8b, v1.8b, v2.8b
cmge v0.16b, v1.16b, v2.16b
cmge v0.4h, v1.4h, v2.4h
cmge v0.8h, v1.8h, v2.8h
cmge v0.2s, v1.2s, v2.2s
cmge v0.4s, v1.4s, v2.4s
sshl v0.8b, v1.8b, v2.8b
sshl v0.16b, v1.16b, v2.16b
sshl v0.4h, v1.4h, v2.4h
sshl v0.8h, v1.8h, v2.8h
sshl v0.2s, v1.2s, v2.2s
sshl v0.4s, v1.4s, v2.4s
sqshl v0.8b, v1.8b, v2.8b
sqshl v0.16b, v1.16b, v2.16b
sqshl v0.4h, v1.4h, v2.4h
sqshl v0.8h, v1.8h, v2.8h
sqshl v0.2s, v1.2s, v2.2s
sqshl v0.4s, v1.4s, v2.4s
srshl v0.8b, v1.8b, v2.8b
srshl v0.16b, v1.16b, v2.16b
srshl v0.4h, v1.4h, v2.4h
srshl v0.8h, v1.8h, v2.8h
srshl v0.2s, v1.2s, v2.2s
srshl v0.4s, v1.4s, v2.4s
sqrshl v0.8b, v1.8b, v2.8b
sqrshl v0.16b, v1.16b, v2.16b
sqrshl v0.4h, v1.4h, v2.4h
sqrshl v0.8h, v1.8h, v2.8h
sqrshl v0.2s, v1.2s, v2.2s
sqrshl v0.4s, v1.4s, v2.4s
smax v0.8b, v1.8b, v2.8b
smax v0.16b, v1.16b, v2.16b
smax v0.4h, v1.4h, v2.4h
smax v0.8h, v1.8h, v2.8h
smax v0.2s, v1.2s, v2.2s
smax v0.4s, v1.4s, v2.4s
smin v0.8b, v1.8b, v2.8b
smin v0.16b, v1.16b, v2.16b
smin v0.4h, v1.4h, v2.4h
smin v0.8h, v1.8h, v2.8h
smin v0.2s, v1.2s, v2.2s
smin v0.4s, v1.4s, v2.4s
sabd v0.8b, v1.8b, v2.8b
sabd v0.16b, v1.16b, v2.16b
sabd v0.4h, v1.4h, v2.4h
sabd v0.8h, v1.8h, v2.8h
sabd v0.2s, v1.2s, v2.2s
sabd v0.4s, v1.4s, v2.4s
saba v0.8b, v1.8b, v2.8b
saba v0.16b, v1.16b, v2.16b
saba v0.4h, v1.4h, v2.4h
saba v0.8h, v1.8h, v2.8h
saba v0.2s, v1.2s, v2.2s
saba v0.4s, v1.4s, v2.4s
add v0.8b, v1.8b, v2.8b
add v0.16b, v1.16b, v2.16b
add v0.4h, v1.4h, v2.4h
add v0.8h, v1.8h, v2.8h
add v0.2s, v1.2s, v2.2s
add v0.4s, v1.4s, v2.4s
cmtst v0.8b, v1.8b, v2.8b
cmtst v0.16b, v1.16b, v2.16b
cmtst v0.4h, v1.4h, v2.4h
cmtst v0.8h, v1.8h, v2.8h
cmtst v0.2s, v1.2s, v2.2s
cmtst v0.4s, v1.4s, v2.4s
mla v0.8b, v1.8b, v2.8b
mla v0.16b, v1.16b, v2.16b
mla v0.4h, v1.4h, v2.4h
mla v0.8h, v1.8h, v2.8h
mla v0.2s, v1.2s, v2.2s
mla v0.4s, v1.4s, v2.4s
mul v0.8b, v1.8b, v2.8b
mul v0.16b, v1.16b, v2.16b
mul v0.4h, v1.4h, v2.4h
mul v0.8h, v1.8h, v2.8h
mul v0.2s, v1.2s, v2.2s
mul v0.4s, v1.4s, v2.4s
smaxp v0.8b, v1.8b, v2.8b
smaxp v0.16b, v1.16b, v2.16b
smaxp v0.4h, v1.4h, v2.4h
smaxp v0.8h, v1.8h, v2.8h
smaxp v0.2s, v1.2s, v2.2s
smaxp v0.4s, v1.4s, v2.4s
sminp v0.8b, v1.8b, v2.8b
sminp v0.16b, v1.16b, v2.16b
sminp v0.4h, v1.4h, v2.4h
sminp v0.8h, v1.8h, v2.8h
sminp v0.2s, v1.2s, v2.2s
sminp v0.4s, v1.4s, v2.4s
sqdmulh v0.4h, v1.4h, v2.4h
sqdmulh v0.8h, v1.8h, v2.8h
sqdmulh v0.2s, v1.2s, v2.2s
sqdmulh v0.4s, v1.4s, v2.4s
addp v0.8b, v1.8b, v2.8b
addp v0.16b, v1.16b, v2.16b
addp v0.4h, v1.4h, v2.4h
addp v0.8h, v1.8h, v2.8h
addp v0.2s, v1.2s, v2.2s
addp v0.4s, v1.4s, v2.4s
uhadd v0.8b, v1.8b, v2.8b
uhadd v0.16b, v1.16b, v2.16b
uhadd v0.4h, v1.4h, v2.4h
uhadd v0.8h, v1.8h, v2.8h
uhadd v0.2s, v1.2s, v2.2s
uhadd v0.4s, v1.4s, v2.4s
uqadd v0.8b, v1.8b, v2.8b
uqadd v0.16b, v1.16b, v2.16b
uqadd v0.4h, v1.4h, v2.4h
uqadd v0.8h, v1.8h, v2.8h
uqadd v0.2s, v1.2s, v2.2s
uqadd v0.4s, v1.4s, v2.4s
urhadd v0.8b, v1.8b, v2.8b
urhadd v0.16b, v1.16b, v2.16b
urhadd v0.4h, v1.4h, v2.4h
urhadd v0.8h, v1.8h, v2.8h
urhadd v0.2s, v1.2s, v2.2s
urhadd v0.4s, v1.4s, v2.4s
uhsub v0.8b, v1.8b, v2.8b
uhsub v0.16b, v1.16b, v2.16b
uhsub v0.4h, v1.4h, v2.4h
uhsub v0.8h, v1.8h, v2.8h
uhsub v0.2s, v1.2s, v2.2s
uhsub v0.4s, v1.4s, v2.4s
uqsub v0.8b, v1.8b, v2.8b
uqsub v0.16b, v1.16b, v2.16b
uqsub v0.4h, v1.4h, v2.4h
uqsub v0.8h, v1.8h, v2.8h
uqsub v0.2s, v1.2s, v2.2s
uqsub v0.4s, v1.4s, v2.4s
cmhi v0.8b, v1.8b, v2.8b
cmhi v0.16b, v1.16b, v2.16b
cmhi v0.4h, v1.4h, v2.4h
cmhi v0.8h, v1.8h, v2.8h
cmhi v0.2s, v1.2s, v2.2s
cmhi v0.4s, v1.4s, v2.4s
cmhs v0.8b, v1.8b, v2.8b
cmhs v0.16b, v1.16b, v2.16b
cmhs v0.4h, v1.4h, v2.4h
cmhs v0.8h, v1.8h, v2.8h
cmhs v0.2s, v1.2s, v2.2s
cmhs v0.4s, v1.4s, v2.4s
ushl v0.8b, v1.8b, v2.8b
ushl v0.16b, v1.16b, v2.16b
ushl v0.4h, v1.4h, v2.4h
ushl v0.8h, v1.8h, v2.8h
ushl v0.2s, v1.2s, v2.2s
ushl v0.4s, v1.4s, v2.4s
uqshl v0.8b, v1.8b, v2.8b
uqshl v0.16b, v1.16b, v2.16b
uqshl v0.4h, v1.4h, v2.4h
uqshl v0.8h, v1.8h, v2.8h
uqshl v0.2s, v1.2s, v2.2s
uqshl v0.4s, v1.4s, v2.4s
urshl v0.8b, v1.8b, v2.8b
urshl v0.16b, v1.16b, v2.16b
urshl v0.4h, v1.4h, v2.4h
urshl v0.8h, v1.8h, v2.8h
urshl v0.2s, v1.2s, v2.2s
urshl v0.4s, v1.4s, v2.4s
uqrshl v0.8b, v1.8b, v2.8b
uqrshl v0.16b, v1.16b, v2.16b
uqrshl v0.4h, v1.4h, v2.4h
uqrshl v0.8h, v1.8h, v2.8h
uqrshl v0.2s, v1.2s, v2.2s
uqrshl v0.4s, v1.4s, v2.4s
umax v0.8b, v1.8b, v2.8b
umax v0.16b, v1.16b, v2.16b
umax v0.4h, v1.4h, v2.4h
umax v0.8h, v1.8h, v2.8h
umax v0.2s, v1.2s, v2.2s
umax v0.4s, v1.4s, v2.4s
umin v0.8b, v1.8b, v2.8b
umin v0.16b, v1.16b, v2.16b
umin v0.4h, v1.4h, v2.4h
umin v0.8h, v1.8h, v2.8h
umin v0.2s, v1.2s, v2.2s
umin v0.4s, v1.4s, v2.4s
uabd v0.8b, v1.8b, v2.8b
uabd v0.16b, v1.16b, v2.16b
uabd v0.4h, v1.4h, v2.4h
uabd v0.8h, v1.8h, v2.8h
uabd v0.2s, v1.2s, v2.2s
uabd v0.4s, v1.4s, v2.4s
uaba v0.8b, v1.8b, v2.8b
uaba v0.16b, v1.16b, v2.16b
uaba v0.4h, v1.4h, v2.4h
uaba v0.8h, v1.8h, v2.8h
uaba v0.2s, v1.2s, v2.2s
uaba v0.4s, v1.4s, v2.4s
sub v0.8b, v1.8b, v2.8b
sub v0.16b, v1.16b, v2.16b
sub v0.4h, v1.4h, v2.4h
sub v0.8h, v1.8h, v2.8h
sub v0.2s, v1.2s, v2.2s
sub v0.4s, v1.4s, v2.4s
cmeq v0.8b, v1.8b, v2.8b
cmeq v0.16b, v1.16b, v2.16b
cmeq v0.4h, v1.4h, v2.4h
cmeq v0.8h, v1.8h, v2.8h
cmeq v0.2s, v1.2s, v2.2s
cmeq v0.4s, v1.4s, v2.4s
mls v0.8b, v1.8b, v2.8b
mls v0.16b, v1.16b, v2.16b
mls v0.4h, v1.4h, v2.4h
mls v0.8h, v1.8h, v2.8h
mls v0.2s, v1.2s, v2.2s
mls v0.4s, v1.4s, v2.4s
pmul v0.8b, v1.8b, v2.8b
pmul v0.16b, v1.16b, v2.16b
umaxp v0.8b, v1.8b, v2.8b
umaxp v0.16b, v1.16b, v2.16b
umaxp v0.4h, v1.4h, v2.4h
umaxp v0.8h, v1.8h, v2.8h
umaxp v0.2s, v1.2s, v2.2s
umaxp v0.4s, v1.4s, v2.4s
uminp v0.8b, v1.8b, v2.8b
uminp v0.16b, v1.16b, v2.16b
uminp v0.4h, v1.4h, v2.4h
uminp v0.8h, v1.8h, v2.8h
uminp v0.2s, v1.2s, v2.2s
uminp v0.4s, v1.4s, v2.4s
sqrdmulh v0.4h, v1.4h, v2.4h
sqrdmulh v0.8h, v1.8h, v2.8h
sqrdmulh v0.2s, v1.2s, v2.2s
sqrdmulh v0.4s, v1.4s, v2.4s
fmaxnm v0.2s, v1.2s, v2.2s
fmaxnm v0.4s, v1.4s, v2.4s
fmaxnm v0.2d, v1.2d, v2.2d
fmla v0.2s, v1.2s, v2.2s
fmla v0.4s, v1.4s, v2.4s
fmla v0.2d, v1.2d, v2.2d
fadd v0.2s, v1.2s, v2.2s
fadd v0.4s, v1.4s, v2.4s
fadd v0.2d, v1.2d, v2.2d
fmulx v0.2s, v1.2s, v2.2s
fmulx v0.4s, v1.4s, v2.4s
fmulx v0.2d, v1.2d, v2.2d
fcmeq v0.2s, v1.2s, v2.2s
fcmeq v0.4s, v1.4s, v2.4s
fcmeq v0.2d, v1.2d, v2.2d
fmax v0.2s, v1.2s, v2.2s
fmax v0.4s, v1.4s, v2.4s
fmax v0.2d, v1.2d, v2.2d
frecps v0.2s, v1.2s, v2.2s
frecps v0.4s, v1.4s, v2.4s
frecps v0.2d, v1.2d, v2.2d
fminnm v0.2s, v1.2s, v2.2s
fminnm v0.4s, v1.4s, v2.4s
fminnm v0.2d, v1.2d, v2.2d
fmls v0.2s, v1.2s, v2.2s
fmls v0.4s, v1.4s, v2.4s
fmls v0.2d, v1.2d, v2.2d
fsub v0.2s, v1.2s, v2.2s
fsub v0.4s, v1.4s, v2.4s
fsub v0.2d, v1.2d, v2.2d
fmin v0.2s, v1.2s, v2.2s
fmin v0.4s, v1.4s, v2.4s
fmin v0.2d, v1.2d, v2.2d
frsqrts v0.2s, v1.2s, v2.2s
frsqrts v0.4s, v1.4s, v2.4s
frsqrts v0.2d, v1.2d, v2.2d
fmaxnmp v0.2s, v1.2s, v2.2s
fmaxnmp v0.4s, v1.4s, v2.4s
fmaxnmp v0.2d, v1.2d, v2.2d
faddp v0.2s, v1.2s, v2.2s
faddp v0.4s, v1.4s, v2.4s
faddp v0.2d, v1.2d, v2.2d
fmul v0.2s, v1.2s, v2.2s
fmul v0.4s, v1.4s, v2.4s
fmul v0.2d, v1.2d, v2.2d
fcmge v0.2s, v1.2s, v2.2s
fcmge v0.4s, v1.4s, v2.4s
fcmge v0.2d, v1.2d, v2.2d
facge v0.2s, v1.2s, v2.2s
facge v0.4s, v1.4s, v2.4s
facge v0.2d, v1.2d, v2.2d
fmaxp v0.2s, v1.2s, v2.2s
fmaxp v0.4s, v1.4s, v2.4s
fmaxp v0.2d, v1.2d, v2.2d
fdiv v0.2s, v1.2s, v2.2s
fdiv v0.4s, v1.4s, v2.4s
fdiv v0.2d, v1.2d, v2.2d
fminnmp v0.2s, v1.2s, v2.2s
fminnmp v0.4s, v1.4s, v2.4s
fminnmp v0.2d, v1.2d, v2.2d
fabd v0.2s, v1.2s, v2.2s
fabd v0.4s, v1.4s, v2.4s
fabd v0.2d, v1.2d, v2.2d
fcmgt v0.2s, v1.2s, v2.2s
fcmgt v0.4s, v1.4s, v2.4s
fcmgt v0.2d, v1.2d, v2.2d
facgt v0.2s, v1.2s, v2.2s
facgt v0.4s, v1.4s, v2.4s
facgt v0.2d, v1.2d, v2.2d
fminp v0.2s, v1.2s, v2.2s
fminp v0.4s, v1.4s, v2.4s
fminp v0.2d, v1.2d, v2.2d
and v0.16b, v1.16b, v2.16b
bic v0.16b, v1.16b, v2.16b
orr v0.16b, v1.16b, v2.16b
eor v0.8b, v1.8b, v2.8b
dup v0.8b, w0
dup v0.2d, x0
dup v0.4h, v1.h[0]
dup v0.4s, v1.s[1]
mov w0, v1.s[2]
mov v0.s[1], v2.s[3]
mov v0.s[1], w2
fcmeq v0.4s, v1.4s, #0.0
fneg v0.4s, v1.4s
